File: python_project/hotel_repository.py
# hotel_repository.py
from pathlib import Path

from hotel import Hotel
from multapartment import Multapartment
from singlapartment import Singlapartment


class HotelRepository:
    def __init__(self, file_path: str | Path):
        self.file = Path(file_path)

    def read_from_file(self) -> Hotel:
        hotel = Hotel()
        if not self.file.exists():
            return hotel
        with self.file.open("r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                parts = line.split()
                # expected: number time [rooms]
                if len(parts) < 2:
                    continue
                num = int(parts[0])
                time = float(parts[1])
                rooms = int(parts[2]) if len(parts) >= 3 else 0
                if rooms:
                    hotel.add_apartment(Multapartment(num, time, rooms))
                else:
                    hotel.add_apartment(Singlapartment(num, time))
        return hotel

    def rewrite_file(self, hotel: Hotel) -> None:
        with self.file.open("w", encoding="utf-8") as f:
            for apt in hotel.output_apartments():
                if apt.is_single():
                    f.write(f"{apt.number} {apt.cleaning_time_per_room}\n")
                else:
                    f.write(f"{apt.number} {apt.get_time()} {apt.room_amount}\n")

File: cpp/hotelrepository.cpp
#include "hotelrepository.h"
#include "multapartment.h"
#include "singlapartment.h"
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

Hotel HotelRepository::readFromFile() const {
    Hotel hotel;
    ifstream fin(file_name);
    if (!fin) {
        cerr << "Unable to open file for reading: " << file_name << endl;
        return hotel;
    }

    string line;
    while (getline(fin, line)) {
        // skip empty / whitespace-only lines
        if (line.find_first_not_of(" \t\r\n") == string::npos) continue;

        istringstream iss(line);
        int room;
        double time;
        int amount;

        if (!(iss >> room >> time)) {
            // malformed line: ignore or log
            cerr << "Skipping malformed line: " << line << endl;
            continue;
        }

        if (iss >> amount) {
            // line had three tokens -> multapartment
            hotel.addApartment(make_unique<Multapartment>(room, time, amount));
        } else {
            // only two tokens -> single-room apartment
            hotel.addApartment(make_unique<Singlapartment>(room, time));
        }
    }

    return hotel;
}


void HotelRepository::rewriteFile(const Hotel& hotel) const {
    ofstream ofs(file_name, ios::trunc);
    if (!ofs) {
        cerr << "Unable to open file for writing: " << file_name << endl;
        return;
    }

    for (Apartment* room : hotel.outputApartments()) {
        ofs << room->getNumber() << " " << room->getTime();
        if (!room->isSingle()) {
            // getRoomAmount exists in base now
            ofs << " " << room->getRoomAmount();
        }
        ofs << endl;
    }
}
